include/domain_bridge/domain_bridge.hpp

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021, Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DOMAIN_BRIDGE__DOMAIN_BRIDGE_HPP_
+#define DOMAIN_BRIDGE__DOMAIN_BRIDGE_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rclcpp/executor.hpp"
+
+#include "domain_bridge/domain_bridge_options.hpp"
+#include "domain_bridge/topic_bridge_options.hpp"
+#include "domain_bridge/visibility_control.hpp"
+
+namespace domain_bridge
+{
+
+class DomainBridgeImpl;
+
+/// Bridge ROS entities across different domains.
+class DomainBridge
+{
+public:
+  /// Constructor.
+  DOMAIN_BRIDGE_PUBLIC
+  explicit DomainBridge(const DomainBridgeOptions & options = DomainBridgeOptions());
+
+  /// Destructor.
+  DOMAIN_BRIDGE_PUBLIC
+  ~DomainBridge();
+
+  DOMAIN_BRIDGE_PUBLIC
+  /// Get the options for this domain bridge.
+  DomainBridgeOptions
+  get_domain_bridge_options() const;
+
+  /// Add the bridge to an executor.
+  /**
+   * When the executor is running ROS traffic will be bridged between domains.
+   * Only topics registered with \ref bridge_topic() will have their messages forwarded.
+   *
+   * This method will add all nodes associated with this bridge to the provided executor.
+   *
+   * \param executor: The executor to add this domain bridge to.
+   */
+  DOMAIN_BRIDGE_PUBLIC
+  void add_to_executor(rclcpp::Executor & executor);
```

Comment:
It's certainly not required to be this way.

The first draft contained it's own executor as an implementation detail, but I refactored after this comment: https://github.com/ros2/domain_bridge/pull/1#discussion_r583707497

I think this API makes the domain_bridge more flexible as a library component of a larger executable. Basically, an executor is needed by the bridge to process subscription callbacks, but choosing the type of executor seems arbitrary.
This API gives users more control over the execution model and how the library is integrated; it lets them tailor the bridge for their application. Compare this with the alternative of providing a single "spin" method, where users are overly constrained.

For users that don't care about choosing an executor, they can use the executable instead, where we've arbitrarily chosen a single threaded executor.

---

I can certainly better document the programming model in the API docs. I hadn't considered documenting the it in the design doc, which is a bit more high-level, but I'm happy to add a section there too if you think it would help.

Reply:
I've updated the API docs a little bit: 48bf2f1